dotnet-engine/Yggdrasil.Engine/NativeLoader.cs

```diff
@@ -0,0 +1,179 @@
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+internal static class NativeLibLoader
+{
+    internal static IntPtr LoadNativeLibrary()
+    {
+        var libName = GetBinaryName();
+        var tempPath = Path.Combine(Path.GetTempPath(), libName);
+        var assembly = Assembly.GetExecutingAssembly();
+        var assemblyName = assembly.GetName().Name;
+
+        if (!File.Exists(tempPath))
+        {
+            using (var stream = Assembly.GetExecutingAssembly().GetManifestResourceStream($"{assemblyName}.{libName}"))
+            {
+                if (stream == null)
+                    throw new FileNotFoundException($"Embedded resource {libName} not found.");
+
+                using (var fileStream = new FileStream(tempPath, FileMode.Create, FileAccess.Write))
+                {
+                    stream.CopyTo(fileStream);
+                }
+            }
+        }
+
+        return LoadBinary(tempPath);
+    }
+
+    private static string GetBinaryName()
+    {
+        string os, arch, libc = "";
+
+        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            os = "win";
+        else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            os = "linux";
+        else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            os = "osx";
+        else
+            throw new PlatformNotSupportedException("Unsupported OS");
+
+        if (RuntimeInformation.ProcessArchitecture == Architecture.X64)
+            arch = "x86_64";
+        else if (RuntimeInformation.ProcessArchitecture == Architecture.Arm64)
+            arch = "arm64";
+        else if (RuntimeInformation.ProcessArchitecture == Architecture.X86)
+            arch = IntPtr.Size == 4 ? "i686" : "x86_64";
+        else
+            throw new PlatformNotSupportedException("Unsupported CPU architecture");
+
+        if (os == "linux" && IsMusl())
+            libc = "-musl";
+
+        string filename = os == "win"
+            ? $"yggdrasilffi_{arch}.dll"
+            : $"libyggdrasilffi_{arch}{libc}.{(os == "osx" ? "dylib" : "so")}";
+
+        return filename;
+    }
+
+    private static bool IsMusl()
+    {
+        if (!RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            return false;
+
+        try
+        {
+            string output = File.ReadAllText("/proc/self/maps");
```

Comment:
Trick stolen from the Java SDK, if the CLR runs on a Linux system and that system is a MUSL system then we must have loaded the MUSL binaries to load the CLR, so the process space must contain something for MUSL here